src/codeview.js

```diff
@@ -329,6 +335,58 @@ var Codeview = GObject.registerClass({
         return GLib.SOURCE_REMOVE;
     }
 
+    // returns a diff in the form of an ed-script, which is the format output by
+    // the diff tool that takes the fewest bytes. It is forward-only; you cannot
+    // apply it in reverse.
+    async _diffFromLastCompiledText() {
+        const oldBytes = ByteArray.fromString(this._lastCompiledText);
+        const oldTextFd = HackToolbox.open_bytes(oldBytes);
+
+        const launcher = new Gio.SubprocessLauncher({
+            flags: Gio.SubprocessFlags.STDIN_PIPE | Gio.SubprocessFlags.STDOUT_PIPE,
+        });
+        launcher.take_fd(oldTextFd, oldTextFd);
+        const proc = launcher.spawnv(['diff', '-e', `/dev/fd/${oldTextFd}`, '-']);
+        const [stdout] = await proc.communicate_utf8_async(this._buffer.text, null);
+        return stdout;
+    }
+
+    async _recordErrorMetrics(results) {
+        if (this._numErrors === 0 && !this._inMetricsEvent)
+            return;
+
+        const recorder = EosMetrics.EventRecorder.get_default();
+        // This will contain info about the user function being edited, when we
+        // gain that feature. Empty string signifies the regular code panel.
+        const userFunction = '';
+        const eventKey = new GLib.Variant('(ss)',
+            [Gio.Application.get_default().applicationId, userFunction]);
+
+        if (this._numErrors && !this._inMetricsEvent) {
+            const payload = new GLib.Variant('(sssa(suquq))', [
+                Gio.Application.get_default().applicationId,
+                userFunction,
+                this._buffer.text,
+                results.map(({start, end, message}) => [message, start.line,
+                    start.column, end.line, end.column]),
```

Comment:
end could be undefined like in Fizzics
```
(com.endlessm.HackToolbox:3): Gjs-WARNING **: 13:47:28.237: JS ERROR: Error while recording error metrics: TypeError: end is undefined
_recordErrorMetrics/payload<@resource:///com/endlessm/HackToolbox/js/codeview.js:371:35
_recordErrorMetrics@resource:///com/endlessm/HackToolbox/js/codeview.js:370:17
setCompileResults@resource:///com/endlessm/HackToolbox/js/codeview.js:423:9
_compile@resource:///com/endlessm/HackToolbox/js/Fizzics/level2.js:194:13
_init/<@resource:///com/endlessm/HackToolbox/js/Fizzics/level2.js:30:13
compile@resource:///com/endlessm/HackToolbox/js/codeview.js:333:9
main@resource:///com/endlessm/HackToolbox/js/main.js:30:12
run@resource:///org/gnome/gjs/modules/package.js:225:12
@/app/bin/com.endlessm.HackToolbox:14:1
```

so we could do something like this
```
                results.map(({start, end, message}) => {
                    end = end || start;
                    return [message, start.line, start.column, end.line, end.column];
                }),
```
or make the last two parameters optionals